include/util/cell_storage.hpp

```diff
@@ -0,0 +1,295 @@
+                    BOOST_ASSERT(std::distance(begin, end) > 0);
+
+                    const auto cell_id = begin->first;
+                    BOOST_ASSERT(level_offset + cell_id < cells.size());
+                    auto &cell = cells[level_offset + cell_id];
+                    cell.num_source_nodes = std::distance(begin, end);
+                    cell.source_boundary_offset = source_boundary.size();
+
+                    std::transform(begin,
+                                   end,
+                                   std::back_inserter(source_boundary),
+                                   [](const auto &cell_and_node) { return cell_and_node.second; });
+                });
+
+            util::for_each_range(
+                level_destination_boundary.begin(),
+                level_destination_boundary.end(),
+                [&](auto begin, auto end) {
+                    BOOST_ASSERT(std::distance(begin, end) > 0);
+
+                    const auto cell_id = begin->first;
+                    BOOST_ASSERT(level_offset + cell_id < cells.size());
+                    auto &cell = cells[level_offset + cell_id];
+                    cell.num_destination_nodes = std::distance(begin, end);
+                    cell.destination_boundary_offset = destination_boundary.size();
+
+                    std::transform(begin,
+                                   end,
+                                   std::back_inserter(destination_boundary),
+                                   [](const auto &cell_and_node) { return cell_and_node.second; });
+                });
+        }
+
+        // Set weight offsets and calculate total storage size
+        WeightOffset weight_offset = 0;
+        for (auto &cell : cells)
+        {
+            cell.weight_offset = weight_offset;
+            weight_offset += cell.num_source_nodes * cell.num_destination_nodes;
+        }
+
+        weights.resize(weight_offset+1, INVALID_EDGE_WEIGHT);
+    }
+
+    CellStorage(std::vector<EdgeWeight> weights_,
+                std::vector<NodeID> source_boundary_,
+                std::vector<NodeID> destination_boundary_,
+                std::vector<CellData> cells_,
+                std::vector<std::size_t> level_to_cell_offset_)
+        : weights(std::move(weights_)), source_boundary(std::move(source_boundary_)),
+          destination_boundary(std::move(destination_boundary_)), cells(std::move(cells_)),
+          level_to_cell_offset(std::move(level_to_cell_offset_))
+    {
+    }
+
+    ConstCell GetCell(LevelID level, CellID id) const
+    {
+        auto offset = level_to_cell_offset[LevelIDToIndex(level)];
+        return ConstCell{
+            cells[offset + id], &weights[0], &source_boundary[0], &destination_boundary[0]};
+    }
+
+    Cell GetCell(LevelID level, CellID id)
+    {
+        auto offset = level_to_cell_offset[LevelIDToIndex(level)];
+        return Cell{cells[offset + id], &weights[0], &source_boundary[0], &destination_boundary[0]};
+    }
```

Comment:
Would love some assertions in the file - especially since you subscript tons of vectors.